pkg/service_sync/service_sync.go

```diff
@@ -0,0 +1,266 @@
+		localServices:     map[string]types.ServiceInterface{},
+		byName:            map[string]types.ServiceInterface{},
+		heardFrom:         map[string]time.Time{},
+	}
+	return s
+}
+
+func (c *ServiceSync) LocalDefinitionsUpdated(definitions map[string]types.ServiceInterface) {
+	c.updates <- definitions
+}
+
+func (c *ServiceSync) pareByOrigin(service string) {
+	for _, origin := range c.byOrigin {
+		if _, ok := origin[service]; ok {
+			delete(origin, service)
+			return
+		}
+	}
+}
+
+func (c *ServiceSync) recordChanges(definitions map[string]types.ServiceInterface) {
+	var added []types.ServiceInterface
+	var modified []types.ServiceInterface
+	var removed []types.ServiceInterface
+
+	for _, def := range c.byName {
+		latest, ok := definitions[def.Address]
+		if !ok {
+			removed = append(removed, def)
+		} else if !equivalentServiceDefinition(&def, &latest) {
+			modified = append(modified, def)
+		}
+	}
+	for _, def := range definitions {
+		if _, ok := c.byName[def.Address]; !ok {
+			added = append(added, def)
+		}
+	}
+	if len(added) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) added %s", strings.Join(getAddresses(added), ","))
+	}
+	if len(removed) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) removed %s", strings.Join(getAddresses(removed), ","))
+	}
+	if len(modified) > 0 {
+		event.Recordf(ServiceSyncEvent, "Service interface(s) modified %s", strings.Join(getAddresses(modified), ","))
+	}
+}
+
+func (c *ServiceSync) localDefinitionsUpdated(definitions map[string]types.ServiceInterface) ServiceUpdate {
+	latest := make(map[string]types.ServiceInterface) // becomes c.localServices
+	byName := make(map[string]types.ServiceInterface)
+
+	for name, original := range definitions {
+		service := types.ServiceInterface{
+			Address:      original.Address,
+			Protocol:     original.Protocol,
+			Ports:        original.Ports,
+			Origin:       original.Origin,
+			Headless:     original.Headless,
+			Labels:       original.Labels,
+			Aggregate:    original.Aggregate,
+			EventChannel: original.EventChannel,
+			Targets:      []types.ServiceInterfaceTarget{},
+		}
+		if service.Origin != "" && service.Origin != "annotation" {
+			if _, ok := c.byOrigin[service.Origin]; !ok {
+				c.byOrigin[service.Origin] = make(map[string]types.ServiceInterface)
+			}
+			c.byOrigin[service.Origin][name] = service
+		} else {
+			latest[service.Address] = service
+			// may have previously been tracked by origin
+			c.pareByOrigin(service.Address)
+		}
+		byName[service.Address] = service
+	}
+
+	c.recordChanges(definitions)
+	c.localServices = latest
+	c.byName = byName
+
+	return ServiceUpdate{
+		origin:      c.origin,
+		version:     c.version,
+		definitions: latest,
+	}
+}
+
+func (c *ServiceSync) updateRemoteDefinitions(origin string, serviceInterfaceDefs map[string]types.ServiceInterface) {
+	event.Recordf(ServiceSyncEvent, "Received update from site %s", origin)
```

Comment:
This is going to generate lots of events. Is it desired or should it only generate an event
when something has changed?

Reply:
It was intentional, but it may not be correct. My thinking was (and this came from debugging a bug), that it gives a good indication that communication is still active. Events from the same site will all be coalesced. It is true that for networks with several sites, these updates will quickly obliterate others however. So perhaps a better solution to ensuring communication is occurring is needed.

Reply:
I've removed that for now. Updates are still recorded (and this patch fixes them so that it is not just changes to services with local targets that are reported).